src/Payments.sol

```diff
@@ -296,86 +434,122 @@ contract Payments is
+        validateAndModifyRateChangeApproval(
+            rail,
+            approval,
+            oldRate,
+            newRate,
+            oneTimePayment,
+            effectiveLockupPeriod
+        );
+
+        // Update the rail fixed lockup and payment rate
         rail.lockupFixed = rail.lockupFixed - oneTimePayment;
         rail.paymentRate = newRate;
 
-        require(
-            payer.lockupRate >= oldRate,
-            "payer lockup rate cannot be less than old rate"
-        );
-        payer.lockupRate = payer.lockupRate - oldRate + newRate;
+        // Update payer's lockup rate - only if the rail is not terminated
+        // for terminated rails, the payer's lockup rate is already updated during rail termination
+        if (!isRailTerminated(rail)) {
+            require(
+                payer.lockupRate >= oldRate,
+                "payer lockup rate cannot be less than old rate"
+            );
+            payer.lockupRate = payer.lockupRate - oldRate + newRate;
+        }
 
-        require(
-            payer.lockupCurrent >=
-                ((oldRate * rail.lockupPeriod) + oneTimePayment),
-            "failed to modify rail payment: insufficient current lockup"
-        );
+        // Update payer's current lockup with effective lockup period calculation
+        // Remove old rate lockup for the effective period, add new rate lockup for the same period
         payer.lockupCurrent =
             payer.lockupCurrent -
-            (oldRate * rail.lockupPeriod) +
-            (newRate * rail.lockupPeriod) -
+            (oldRate * effectiveLockupPeriod) +
+            (newRate * effectiveLockupPeriod) -
             oneTimePayment;
 
         // --- Process the One-Time Payment ---
         processOneTimePayment(payer, payee, oneTimePayment);
 
+        // Ensure the modified lockup doesn't exceed available funds
         require(
             payer.lockupCurrent <= payer.funds,
-            "payer lockup cannot exceed funds"
+            "invariant violation: payer lockup cannot exceed funds"
         );
+
+        // If we've reduced the rate, settle lockup again to account for changes
+        // and ensure that account lockup is settled upto and including the current epoch
+        if (newRate < oldRate) {
+            uint256 settledUpto = settleAccountLockup(payer);
+            require(
+                settledUpto == block.number,
+                "account lockup must be fully settled after rate decrease"
+            );
+        }
+    }
+
+    function handleRateChangeSettlement(
+        uint256 railId,
+        Rail storage rail,
+        uint256 oldRate,
+        uint256 newRate
+    ) internal {
+        // If rate hasn't changed, nothing to do
+        if (newRate == oldRate) {
+            return;
+        }
+
+        // If there is no arbiter, settle the rail immediately
+        if (rail.arbiter == address(0)) {
+            (, uint256 settledUpto, ) = settleRail(railId, block.number, false);
```

Comment:
change this to take a rail?

Reply:
(trying to avoid passing rail + rail id everywhere)

Reply:
This is the only place we pass a `railId` now because `settleRail` is a public function that can be called by EOAs as well. And they only need to pass the `railId`.

Reply:
Ah, I see.